src/features/todo/todoSlice.jsx

```diff
@@ -0,0 +1,30 @@
+import { createSlice } from '@reduxjs/toolkit'
+
+const initialState = {
+    products: JSON.parse(localStorage.getItem('products')) || []
+}
+
+export const productSlice = createSlice({
+    name: 'products',
+    initialState,
+    reducers: {
+        addProduct: (state, action) => {
+            let found = false;
+            state.products.forEach((product) => {
+                if (product.info.name === action.payload.info.name) {
+                    product.number = (parseInt(product.number) || 0) + parseInt(action.payload.number);
+                    found = true;
+                }
+            })
+            if (!found) { state.products.push(action.payload) }
+            localStorage.setItem('products', JSON.stringify(state.products));
```

Comment:
reducers should be [pure functions](https://www.geeksforgeeks.org/pure-functions-in-javascript/), remove all side effects from it.
read again about [reducers and how they should work in the docs](https://redux.js.org/tutorials/fundamentals/part-3-state-actions-reducers#writing-reducers), they should return new state. and your reducers return nothing. fix here and everywhere else